tornado/httpserver.py

```diff
@@ -179,12 +179,78 @@ def handle_stream(self, stream, address):
         conn.start_serving(self)
 
     def start_request(self, server_conn, request_conn):
-        return _ServerRequestAdapter(self, server_conn, request_conn)
+        if isinstance(self.request_callback, (tuple, list)):
```

Comment:
I was hoping we'd be able to do this without changes to HTTPServer: the router would implement HTTPServerConnectionDelegate, and it would be the only thing that knows about the (app, request) binding.